oio/xcute/orchestrator.py

```diff
@@ -366,77 +350,104 @@ def process_reply(self, beanstalkd_job_id, encoded_reply):
+                if not self.running or self._refresh_beanstalkd_workers:
+                    break
+                sleep(1)
 
-                self.all_beanstalkd[beanstalkd_addr] = beanstalkd + ({},)
+        self.logger.info('Exited beanstalkd workers thread')
 
-            sleep(5)
-
-        self.logger.info('Exited beanstalkd thread')
+    def _find_beanstalkd_workers(self, beanstalkd_workers,
+                                 beanstalkd_stats):
+        """
+        Find beanstalkd workers by looking at the score,
+        existing tubes and tube statistics.
+        """
+        all_beanstalkd = self.conscience_client.all_services(
+            'beanstalkd')
 
-    @staticmethod
-    def get_beanstalkd_tubes(beanstalkd_addr):
-        return Beanstalk.from_url('beanstalkd://' + beanstalkd_addr).tubes()
+        for beanstalkd_info in all_beanstalkd:
+            try:
+                self._check_beanstalkd_worker(
+                    beanstalkd_workers, beanstalkd_stats, beanstalkd_info)
+            except Exception as exc:
+                self.logger.error('Fail to check beanstalkd: %s', exc)
 
-    def get_loadbalanced_workers(self):
+    def _check_beanstalkd_worker(self, beanstalkd_workers,
+                                 beanstalkd_stats, beanstalkd_info):
         """
-            Yield senders following a loadbalancing strategy
+        Check beanstalkd worker by looking at the score,
+        existing tubes and tube statistics.
         """
+        beanstalkd_score = beanstalkd_info['score']
+        if beanstalkd_score == 0:
+            return
 
-        while True:
-            yielded = False
-            for beanstalkd, beanstalkd_tubes, beanstalkd_senders \
-                    in self.all_beanstalkd.itervalues():
-                if beanstalkd['score'] == 0:
-                    continue
+        beanstalkd_addr = 'beanstalkd://' + beanstalkd_info['addr']
+        beanstalkd = Beanstalk.from_url(beanstalkd_addr)
 
-                if self.beanstalkd_workers_tube not in beanstalkd_tubes:
-                    continue
+        beanstalkd_tubes = beanstalkd.tubes()
+        if self.beanstalkd_workers_tube \
+                not in beanstalkd_tubes:
+            # No worker has ever listened to the tube.
+            return
 
-                if self.beanstalkd_workers_tube not in beanstalkd_senders:
-                    sender = BeanstalkdSender(
-                        addr=beanstalkd['addr'],
-                        tube=self.beanstalkd_workers_tube,
-                        logger=self.logger)
+        current_stats = beanstalkd.stats_tube(
+            self.beanstalkd_workers_tube)
+        previous_stats = beanstalkd_stats.get(beanstalkd_addr)
+        beanstalkd_stats[beanstalkd_addr] = current_stats
+        if (previous_stats is not None
+                and (previous_stats['current-jobs-ready'] > 0
+                     or current_stats['current-jobs-ready'] > 0)
```

Comment:
Is `current-jobs-ready` the number of jobs in the queue?
It's completely normal to have some waiting in the queue (it limits the overhead) and we should strive to always have some in the queue.